sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -110,11 +121,30 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
         "Operator will receive unsafe rows as input but cannot process unsafe rows")
     }
     RDDOperationScope.withScope(sparkContext, nodeName, false, true) {
+      prepare()
       doExecute()
     }
   }
 
   /**
+   * Do the preparation for SparkPlan.
+   */
+  final def prepare(): Unit = {
```

Comment:
can this one just walks down the query plan and call prepare? In that case, we don't need any of the super.doPrepare